### PR TITLE
wr-73: Create Comments Table

### PR DESCRIPTION
**Pull request by b0bbobovich**
- add common comments types for requests and response and minimal vlidation scheme;
- add migration for creating Comment table;
- create repository/service/controller/entity for Comment;
- routes for get comment by id, get all article comments, create, update and delete comment.

### CODE REVIEW

File: backend/src/packages/comments/comment.service.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        ...comment,
+        ...payload,
+      }),
+    );
+
+    return updatedComment.toObject();
+  }
+
+  public async delete(id: number): Promise<boolean> {
+    return await this.commentRepository.delete(id);
```
Comment (artemmatiushenko1) [inline backend/src/packages/comments/comment.service.ts]: I guess there's no need to await promise here, we can simply return it without awaiting

File: frontend/src/assets/img/article-banner.jpg
Comment (artemmatiushenko1) [inline frontend/src/assets/img/article-banner.jpg:1]: I guess it should not be present in this PR, looks like here's the work from the other branch

File: backend/src/db/migrations/20230829161512_create_comments_table.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      .inTable(TableName.ARTICLES);
+    table
+      .dateTime(ColumnName.CREATED_AT)
+      .notNullable()
+      .defaultTo(knex.fn.now());
+    table
+      .dateTime(ColumnName.UPDATED_AT)
+      .notNullable()
+      .defaultTo(knex.fn.now());
+    table.dateTime(ColumnName.PUBLISHED_AT).nullable();
```
Comment (andanf-e) [inline backend/src/db/migrations/20230829161512_create_comments_table.ts]: This column is not needed - comment is published when it is created
```suggestion

```

File: backend/src/packages/comments/comment.model.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import {
+  AbstractModel,
+  DatabaseTableName,
+} from '~/libs/packages/database/database.js';
+
+class CommentModel extends AbstractModel {
+  public 'text': string;
+  public 'userId': number;
+  public 'articleId': number;
+  public 'publishedAt': string | null;
```
Comment (andanf-e) [inline backend/src/packages/comments/comment.model.ts]: According to comment in migration
```suggestion

```

File: shared/src/packages/comments/libs/enums/comment-validation-message.enum.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+const CommentValidationMessage = {
+  COMMENT_TEXT_REQUIRE: 'Comment text is required',
+  COMMENT_PUBLISHED_ISO_DATE:
+    'The date of publication should be in ISO 8601 format',
```
Comment (andanf-e) [inline shared/src/packages/comments/libs/enums/comment-validation-message.enum.ts]: According to comment in migration
```suggestion

```

File: shared/src/packages/comments/libs/validation-schemas/comment-create.validation-schema.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import { type CommentBaseRequestDto } from '~/packages/comments/libs/types/types.js';
+
+const commentCreate = joi.object<CommentBaseRequestDto, true>({
+  text: joi.string().trim().required().messages({
+    'string.empty': CommentValidationMessage.COMMENT_TEXT_REQUIRE,
+  }),
+  articleId: joi.number().integer().positive(),
+  publishedAt: joi.string().isoDate().required().allow(null).messages({
+    'string.isoDate': CommentValidationMessage.COMMENT_PUBLISHED_ISO_DATE,
+  }),
```
Comment (andanf-e) [inline shared/src/packages/comments/libs/validation-schemas/comment-create.validation-schema.ts]: According to comment in migration
```suggestion

```

File: shared/src/packages/comments/libs/validation-schemas/comment-update.validation-schema.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import joi from 'joi';
+
+import { CommentValidationMessage } from '~/packages/comments/libs/enums/enums.js';
+import { type CommentUpdateRequestDto } from '~/packages/comments/libs/types/types.js';
+
+const commentUpdate = joi.object<CommentUpdateRequestDto, true>({
+  text: joi.string().trim(),
+  publishedAt: joi.string().isoDate().allow(null).messages({
+    'string.isoDate': CommentValidationMessage.COMMENT_PUBLISHED_ISO_DATE,
+  }),
```
Comment (andanf-e) [inline shared/src/packages/comments/libs/validation-schemas/comment-update.validation-schema.ts]: Same from comment in migration
```suggestion
 
```

File: shared/src/packages/comments/libs/types/comment-entity.type.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+type CommentEntityType = {
+  id: number;
+  text: string;
+  userId: number;
+  articleId: number;
+  publishedAt: string | null;
```
Comment (andanf-e) [inline shared/src/packages/comments/libs/types/comment-entity.type.ts]: Or leave it with `publishedAt` field but map createdAt column value to it
```suggestion
type CommentEntity = {
  id: number;
  text: string;
  userId: number;
  articleId: number;
  createdAt: string;
```

**Review by andanf-e — CHANGES_REQUESTED** [review]

File: backend/src/db/migrations/20230829161512_create_comments_table.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+} as const;
+
+const ColumnName = {
+  ID: 'id',
+  TEXT: 'text',
+  USER_ID: 'user_id',
+  ARTICLE_ID: 'article_id',
+  CREATED_AT: 'created_at',
+  UPDATED_AT: 'updated_at',
+  PUBLISHED_AT: 'published_at',
```
Comment (andanf-e) [inline backend/src/db/migrations/20230829161512_create_comments_table.ts]:
```suggestion

```

File: backend/src/libs/interfaces/service.interface.ts [hunk 9]
```diff
@@ -5,7 +5,7 @@ interface IService<T = unknown> {
   }>;
   create(payload: unknown): Promise<T>;
   update(id: unknown, payload: unknown): Promise<T>;
-  delete(): Promise<boolean>;
+  delete(id: unknown, payload: unknown): Promise<boolean>;
```
Comment (andanf-e) [inline backend/src/libs/interfaces/service.interface.ts]: 1 - We may not need to delete comments
2 - Default delete should work from id
```suggestion
  delete(): Promise<boolean>;
```

File: backend/src/packages/comments/comment.model.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+class CommentModel extends AbstractModel {
+  public 'text': string;
+  public 'userId': number;
+  public 'articleId': number;
+
+  public static override get tableName(): string {
+    return DatabaseTableName.COMMENTS;
+  }
+}
```
Comment (andanf-e) [inline backend/src/packages/comments/comment.model.ts:37]: Add author relation to userModel

File: backend/src/packages/comments/comment.repository.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    this.commentModel = commentModel;
+  }
+
+  public findAll(): Promise<unknown[]> {
+    return Promise.resolve([]);
+  }
+
+  public async findAllByArticleId(
+    articleId: number,
+  ): Promise<CommentEntity[] | []> {
```
Comment (andanf-e) [inline backend/src/packages/comments/comment.repository.ts]:
```suggestion
  ): Promise<CommentEntity[]> {
```

File: backend/src/packages/comments/comment.service.ts [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    const comment = await this.find(id);
+
+    if (!comment) {
+      throw new ApplicationError({
+        message: `Comment with id ${id} not found`,
+      });
+    }
+
+    return comment;
+  }
```
Comment (andanf-e) [inline backend/src/packages/comments/comment.service.ts]: If it's similar to find method - then why do we need it? 
```suggestion

```

**Review by andanf-e — CHANGES_REQUESTED** [review]

File: backend/src/libs/interfaces/service.interface.ts (hunk 9 above)
Comment (canterbery) [reply backend/src/libs/interfaces/service.interface.ts]: Can you clarify second point please ?

Comment (andanf-e) [reply backend/src/libs/interfaces/service.interface.ts]: Andrew've added to common interface args with payload on `delete` method -- common delete method should get only `id` for deletion in general

Comment (canterbery) [reply backend/src/libs/interfaces/service.interface.ts]: Yes, I saw those changes and delete methods in comment repository and service. I just don't undestand what I need to do - rewrite it using only id or remove entirely ?

Comment (andanf-e) [reply backend/src/libs/interfaces/service.interface.ts]: Just remove changes

File: shared/src/packages/comments/libs/validation-schemas/comment-update.validation-schema.ts [hunk 13]
```diff
@@ -0,0 +1,9 @@
+import joi from 'joi';
+
+import { type CommentUpdateRequestDto } from '~/packages/comments/libs/types/types.js';
+
+const commentUpdate = joi.object<CommentUpdateRequestDto, true>({
+  text: joi.string().trim(),
```
Comment (drMangust) [inline shared/src/packages/comments/libs/validation-schemas/comment-update.validation-schema.ts]:
```suggestion
  text: joi.string().trim().required().messages({
    'string.empty': CommentValidationMessage.COMMENT_TEXT_REQUIRE,
  }),
```

Comment (drMangust) [reply shared/src/packages/comments/libs/validation-schemas/comment-update.validation-schema.ts]: Perhaps we should make the "text" field also mandatory for the schema?

**Review by andanf-e — DISMISSED** [review]

**Review by andanf-e — APPROVED** [review]

**Review by mhrechkin — APPROVED** [review]

**Pull request merged**